_episodes/02-generalsyntax.md

```diff
@@ -0,0 +1,264 @@
+---
+title: "General Syntax of a Shell Command"
+teaching: ??
+exercises: ??
+questions:
+- "How do I tell the shell to do something?"
+- "How do flags and options modify a shell program?"
+- "How do I find more information about specific shell commands?"
+objectives:
+- "Explain the correct general format of a shell command."
+- "Explain man pages and the --help flag."
+- "Explain how to supply options/flags to a shell command."
+keypoints:
+- "The basic format of a shell command is <command> <flags/options> <arguments>."
+- "More information about commands and their flags can be found in the manual, or 'man' pages."
+- "The --help option is also useful for learning more about commands."
+- "Some commands have default arguments."
+- "Man pages are also available on the internet."
+---
+## General syntax of a shell command
+Consider the command below as a general example of a command,
+which we will dissect into its component parts:
+
+~~~
+$ ls -F /
+~~~
+{: .language-bash}
+
+
+`ls` is the **command**, with an **option** `-F` and an
+**argument** `/`. Options (also called **switches** or **flags**)
+either start with a single dash (`-`) or two dashes (`--`), and they change the behaviour of a command.
+Arguments tell the command what to operate on (e.g. files and directories).
+Sometimes options and arguments are referred to as **parameters**.
+A command can be called with more than one option and more than one argument, but a
+command doesn't always require an argument or an option.
+
+Each part is separated by spaces: if you omit the space
+between `ls` and `-F` the shell will look for a command called `ls-F`, which
+doesn't exist. Also, capitalization can be important: `ls -r` is different from `ls -R`.
+
+Putting all that together, our command above gives us a listing
+of files and directories in the root directory `/`.
+An example of the output you might get from the above command is given below:
+
+~~~
+$ ls -F /
+Applications/         System/
+Library/              Users/
+Network/              Volumes/
+~~~
+{: .output}
+
+> ## Default Flags
+>
+> Some commands, when run without any flags, will assume one or more by default.
+> A good example of this is the 'wc' command, which has the default flags -c, -l,
+> and -w. This means that running 'wc' is equivalent to running 'wc -clw'. If a
+> command has default flags, they will be listed in the manual, or 'man' page, usually under
+> the 'Description' section. We will cover man pages and how to read them in the
```

Comment:
Default options for `wc` are not listed anywhere in my man pages or `--help` output on Linux Mint. Which OS are you using?

Reply:
I'm using Mac OSX. The part I'm referring to is at the bottom of the Description section (see attached screenshot) that reads 'The default action is equivalent to specifying the -c, -l and -w options." 

<img width="653" alt="Screen Shot 2019-07-02 at 9 08 18 AM" src="https://user-images.githubusercontent.com/5347501/60511711-fb88dc00-9ca8-11e9-91b1-d61b948eb4bc.png">

Reply:
Ah, I see. This isn't given in my man pages on Linux. We should be trying to avoid material where the output is very different depending on platform. I don't know what the man pages look like on Git Bash.

```
SYNOPSIS
       wc [OPTION]... [FILE]...
       wc [OPTION]... --files0-from=F

DESCRIPTION
       Print  newline,  word,  and byte counts for each FILE, and a total line if more than one FILE is specified.  A
       word is a non-zero-length sequence of characters delimited by white space.

       With no FILE, or when FILE is -, read standard input.

       The options below may be used to select which counts are printed, always  in  the  following  order:  newline,
       word, character, byte, maximum line length.

       -c, --bytes
              print the byte counts

       -m, --chars
              print the character counts

       -l, --lines
              print the newline counts

       --files0-from=F
              read  input  from the files specified by NUL-terminated names in file F; If F is - then read names from
              standard input

       -L, --max-line-length
              print the maximum display width

       -w, --words
              print the word counts

       --help display this help and exit

       --version
              output version information and exit

AUTHOR
       Written by Paul Rubin and David MacKenzie.

REPORTING BUGS
       GNU coreutils online help: <http://www.gnu.org/software/coreutils/>
       Report wc translation bugs to <http://translationproject.org/team/>

COPYRIGHT
       Copyright  ©  2016  Free  Software  Foundation,  Inc.   License  GPLv3+:  GNU   GPL   version   3   or   later
       <http://gnu.org/licenses/gpl.html>.
```

Reply:
Implicit on the manpage, but its there:
> Print  newline,  word,  and byte counts for each FILE